Port Zeek to latest Broker API (ALM Transport Branch)

Neverlord opened the pull request:
Zeek-side adjustments for https://github.com/zeek/broker/pull/89.

Neverlord commented:
Most errors in Zeek seem to come from the new Broker status events. I'll look into it in more detail next week.

Neverlord commented:
Reverting back to draft for now, since I'll have to update several btest baselines, BIFs, and docs.

Neverlord commented:
Aside from straightforward updates caused by changes to Broker/CAF API or log output, there were also a couple of subtle shifts in semantics worth pointing out:

- After switching to an overlay transport, Broker now distinguishes between *known endpoints* and *peers*. A known endpoint is a Broker endpoint that's currently active in the network and reachable. A peer is a Broker endpoint that's reachable in one hop, i.e., through a direct network connection. Hence, there are two new Zeek events: `Broker::endpoint_discovered` and `Broker::endpoint_unreachable`. When peering to a previously unknown endpoint, Broker triggers `Broker::endpoint_discovered` first and then `Broker::peer_added`. I hope I've added docs in all the right places.
- When failing to connect to an endpoint via `peer`, Broker now triggers `peer_unavailable` *once* after the final connection attempt failed. When calling `peer`, users can now expect to see either a `peer_added` event *or* a `peer_unavailable` event. Previously, Broker triggered `peer_unavailable` for each failed attempt individually. This means users could observe up to ten "results" for a single call to `peer` (nine unavailable errors plus `peer_added` at the end). This made it virtually impossible for a user to know whether `peer` actually failed when seeing `peer_unavailable`, because Broker could still eventually succeed (and the number of attempts is a hardcoded implementation detail). The one-to-one correlation allows Zeek scripts to come to accurate conclusions now. If we still want some form of feedback to the user on failed attempts, I think we should instead introduce a new status event like `failed_connection_attempt`. For testing purposes, this change means that the OpenSSL auth test now has to wait ten seconds (1s retry interval, 10 attempts) until Broker "gives up". Technically, Broker could abort earlier for this particular case, since we can pretty much assume that a failed OpenSSL handshake means that we can try all we want but always get the same error. Same for incompatible Broker versions. That would require changes to both CAF and Broker though, so if this is something we'd want, I think we should track this with a separate Broker issue. Also, I think it would be a nice addition to the API if we'd make the number of connection attempts configurable by the user.

I've also put all btests for Broker into a group to make it easier to run these without running the entire suite.

jsiwek commented:
Thanks for the update.

Was the the hardcoded maximum limit of 10 connection attempts the old behavior or something new in the branch?  Maybe my memory is bad, but thought the old behavior was to attempt (re)connections indefinitely until it succeeds.

Regardless, I can also see the possible convenience of extending `peer()` to let users pick a maximum retry-count, with a special value for "infinite".  If the old behavior was indeed "infinite", can stick with that for any new default settings.

Definitely useful to be able to distinguish between a "failed attempt" and a "failed attempt, will never retry".  Rather than change the meaning of the `peer_unavailable` event from the former to the later, may be less jarring and backward compatible to introduce a new event with the later meaning, say `peer_unavailable_limit` or `connection_attempt_limit_reached`.

Neverlord commented:
> Was the the hardcoded maximum limit of 10 connection attempts the old behavior or something new in the branch? Maybe my memory is bad, but thought the old behavior was to attempt (re)connections indefinitely until it succeeds.

No, your memory serves you right. I've just checked, v1.3.3 indeed tries forever: https://github.com/zeek/broker/blob/v1.3.3/src/core_actor.cc#L75 (the delayed_send eventuall calls `try_once` again). I'm not sure when/why it changed to 10, but it's already 10 in the `master` branch. So I've probably slipped this in with the core actor redesign without realizing the change in behavior.

> Regardless, I can also see the possible convenience of extending `peer()` to let users pick a maximum retry-count, with a special value for "infinite". If the old behavior was indeed "infinite", can stick with that for any new default settings.

Ok. Should I implement this as a followup or add it to this PR?

> Definitely useful to be able to distinguish between a "failed attempt" and a "failed attempt, will never retry". Rather than change the meaning of the `peer_unavailable` event from the former to the later, may be less jarring and backward compatible to introduce a new event with the later meaning, say `peer_unavailable_limit` or `connection_attempt_limit_reached`.

Sorry, I think I've mixed a couple of things here. Only `broker::sc` codes trigger Zeek events. In other words, there is no `peer_unavailable` event for Zeek scripts. There's only the `PEER_UNAVAILABLE` error code: https://github.com/zeek/zeek/blob/43e54c7/scripts/base/frameworks/broker/main.zeek#L155. I didn't realize that the errors aren't propagated as Zeek events before.

The docs say about this error code "Remote peer not listening". So I'd argue the behavior when Broker triggers this error code differs from what's documented. To me, this does not read as "temporary error, will retry later". None of the `btest` tests for Broker check for this code in script land and only the tests that assume connection failures check for this error code in their baseline. So it seems like the tests already assume that Broker sends `PEER_UNAVAILABLE` errors when giving up on trying to connect to a remote node. However, *any* of the tests could trigger this error code if the process that calls `Broker::listen` spins up the socket too late for the first `Broker::peer` to succeed right away. Seems to not happen in practice because of the way `btest` launches the Zeek processes. I think the new behavior brings the implementation in line with how the tests assume things are and how the error code is documented.

If we would want to make connection attempts visible to the user, we could add a new status code to Broker. This would then also trigger a new Zeek event that could include how many attempts are left (`0` for giving up and some special value for `infinite`). Maybe that should be a followup, though.

jsiwek commented:
> I'm not sure when/why it changed to 10, but it's already 10 in the `master` branch

`master` behavior is still a bit different than `multi-hop-routing`, with `master` imposing the retry limit when a retry interval of "0 seconds" is requested, but `multi-hop-routing` imposing the retry limit when the retry interval is non-zero.

I'd say `master` behavior is a bit strange since a retry interval of zero is supposed to mean "do not retry", but probably not too noticeable/bad, so would concentrate on getting the `multi-hop-routing` branch behavior like we want.

> > convenience of extending `peer()` to let users pick a maximum retry-count, with a special value for "infinite". If the old behavior was indeed "infinite", can stick with that for any new default settings.
> 
> Ok. Should I implement this as a followup or add it to this PR?

Seems more like a followup to me, but I'm generally not even really asking for it, just saying if you think it's a nice extension to the API, seems cool to do that whenever you want.

I see the minimal requirement for the PR as preserving these behaviors if we can:
* A retry interval of "0 seconds" should mean "never retry"
* For non-zero retry intervals, the default behavior should make an infinite number of retry attempts

> The docs say about this error code "Remote peer not listening". So I'd argue the behavior when Broker triggers this error code differs from what's documented. To me, this does not read as "temporary error, will retry later".

If we're choosing between changing the actual behavior and clarifying the documentation, the later seems more appropriate in this case.

> So it seems like the tests already assume that Broker sends `PEER_UNAVAILABLE` errors when giving up on trying to connect to a remote node

In the historical default Zeek/Broker configuration, there is no "giving up", so tests have not been written with that assumption.  For the couple SSL tests that have `PEER_UNAVAILABLE` in their baseline, they've been explicitly written to terminate their process on seeing the first instance of an error.

> However, _any_ of the tests could trigger this error code if the process that calls `Broker::listen` spins up the socket too late for the first `Broker::peer` to succeed right away. Seems to not happen in practice because of the way `btest` launches the Zeek processes

Yes, that's right that any test could trigger a `PEER_UNAVAILABLE`, but either they need to consider that output irrelevant to what they're testing or else use some form of synchronization to avoid startup/scheduling races (the `ssl_auth_failure.zeek` test as I just glanced at it may actually have that unintended race, which I'll look into later).

Think these situations do happen in practice, but more often in CI jobs.  In those cases, I also want to say it's a helpful indicator to see `PEER_UNAVAILABLE` spam in the logs/output when going in to debug a flaky test since otherwise I might have to make an assumption of what other state a process might be stuck in.

> If we would want to make connection attempts visible to the user, we could add a new status code to Broker. This would then also trigger a new Zeek event that could include how many attempts are left (`0` for giving up and some special value for `infinite`). Maybe that should be a followup, though.

The "connection attempt limit" and "connection attempt limit exceeded" ideas seem like possible followup/convenience features to me.  They're unrelated to what we're trying to do here.

Still think it's worth keeping `PEER_UNAVAILABLE` behavior similar to what it was before if we can.  If the previous documentation is a source of confusion, I'd prefer to clarify the meaning as "failed connection attempt, may retry later according to peer() arguments".  Would the following patch be enough to restore previous behavior in `multi-hop-routing`?  I may still misunderstand things, but at least observed Zeek retrying connections past 10 attempts:

```diff
diff --git a/include/broker/mixin/connector.hh b/include/broker/mixin/connector.hh
index 2e4577d..8eab5c2 100644
--- a/include/broker/mixin/connector.hh
+++ b/include/broker/mixin/connector.hh
@@ -66,9 +66,11 @@ public:
             [=](error& err) mutable { rp.deliver(std::move(err)); });
       },
       [=](error err) mutable {
-        if (addr.retry.count() == 0 || ++count >= 10) {
+        // Note: connection retry count is tracked, but currently unused.
+        ++count;
+        dref().peer_unavailable(addr);
+        if (addr.retry.count() == 0) {
           rp.deliver(std::move(err));
-          dref().peer_unavailable(addr);
         } else {
           self->delayed_send(self, addr.retry,
                              detail::retry_state{addr, std::move(rp), count});
```

Neverlord commented:
Ok, I've reverted the behavior back to how it worked in Broker v1.3.3: infinite retries and "errors" on each failed attempt. I have a hard time believing anyone is actually depending on this behavior for `peer_unavailable`, though, since even the Zeek tests read to me like they assume different semantics. The current `peer_unavailable` is a notification, not an error, and should be implemented as such IMHO. I'm not gonna argue here, just pointing it out.

With the Zeek test suites, I found an issue with the flare actor that has been there for a long time (probably since the flare actor exists). Essentially, any unexpected message to a store proxy immediately sends Zeek into an infinite loop. This got triggered by an accidental change to response messages for `put_unique`. I've fixed both. Unexpected messages now trigger Broker errors instead of blowing up the system. This was nasty to debug, since it blew up in a seemingly unrelated test (`scripts/policy/conn/known-services-multi.zeek`).

Anything else that needs to be taken care of for this PR, except getting all CI checks green?

jsiwek commented:
> The current `peer_unavailable` is a notification, not an error, and should be implemented as such IMHO.

Excluding the case of "0 second retry" meaning of "never retry", agree it seems more like a status notification than an error, but the benefit of a change here doesn't seem great enough to warrant an immediate one rather than a deprecation/migration process.  Even if the reality is that zero users are impacted by an immediate change, reluctant to make a habit of giving the impression of an unstable platform that doesn't welcome them to rely on such things in the first place.

> Anything else that needs to be taken care of for this PR, except getting all CI checks green?

Nope, just waiting to see CI is stable (looks close) before any next review/merge steps, thanks.

Neverlord commented:
> Nope, just waiting to see CI is stable (looks close)

Yeah, looked close... Then I started digging. 😅

After lots of troubleshooting and bugfixing, this is what I've fixed in the meantime (most bugs were actually unrelated to the ALM branch):
- Some tests failed sporadically because Broker 'lost' the last couple of events. This was on me, since I didn't re-implement graceful shutdown properly again after the ALM changes.
- The Broker tests never caught message loss during shutdown, so I've added new tests to put stress on peering on unpeering. They worked so well, they've also revealed a race in the handshake logic.
- The handshake logic was spread over multiple functions and had some code duplication. I've ended up re-implementing the handshake using a new FSM abstraction. The handshake now contains an additional step to fix the race from before (document in the developer section in the Broker manual).
- On rare occasions, Broker could die with a segfault due to a nasty bug in Broker's internal dispatching logic. Also fixed.
- The Broker manager in Zeek had a race condition on the message ordering. Many tests actually rely on ordering of events, for example seeing `peer_added` before receiving data on the topics. However, Zeek used two separate subscriber objects: one for status events and one for published data from peers. By using only a single subscriber object, Zeek now always propagates events to scripts in the order Broker emits them. The two different subscribers actually had historic reasons and some recent changes to Broker enabled using a single subscriber now.
- When handling subscription updates from peers, Broker did not check for stale data correctly. In some cases, "old" messages could override peer filters. Due to the subscription flooding, it's actually not that unlikely to receive such stale messages.
- The `broker::subscriber` also had a bug: when passing it an initial filter in the constructor, the first call to `add_topic` would *override* the initial filter instead of appending to it.
- The `brokerstore-backend-simple-reverse` test relied on a particular ordering of events: clone 1 assumed that it would see `announce_master` *before* `node_up`. Intuitively, I’d assume the exact opposite. It makes more sense to me to see the `node_up` for manager 1 before this node announces its store master. This test failed after my other patches (which now always seem to trigger `node_up` *first*), so I’ve updated the test script to be less reliant on a particular order (see cb0adac900177d1de9a5433e86a336d473d6f9d5). @0xxon you probably know the details on this best: is there a deeper reason why `announce_masters` should arrive before `node_up`? Are the test script changes OK in you opinion?
- Last but not least: why did the FreeBSD build on Cirrus CI keep failing? The exact *why* is still a mystery to me, since I cannot reproduce this on any other FreeBSD installation I could get my hands on (neither natively installed nor virtualized). But here's *how* the tests failed (had to enable CAF trace logs during the CI build to figure that one out). Broker tries to auto-connect to peers when they disconnect. For that, Broker stores network information for each peer. This information gets also stored for connecting peers, but here the TCP port is of course not a listening port but the port used during the connection. However, the ALM branch did not check for this and the "server" peer would try to reconnect to a "client". On any other setup, this connection attempt simply fails and then Broker stops trying to connect to this network address after seeing a retry interval of 0. *Except* on the Cirrus CI FreeBSD build. Here, it just hangs and `btest` kills the process after 30s. I’ve added an additional check to fix this minor bug as well.

Getting trace logs from Broker requires CAF 0.18, so I did some hacky integration during the debugging process (hence the force pushes). I think this is something we definitely should enable by default, would've saved me a lot of time. Broker and Zeek work fine with CAF 0.18 already, only the CMake setup requires changes. I’d rather to this migration as part of a followup, though. The PR is already large enough as-is.

rsmmr commented:
Great work tracking this all down, thanks for going through that pain!

"looks close" now I would say. 😉

jsiwek commented:
Thanks for doing all that digging.

> * Last but not least: why did the FreeBSD build on Cirrus CI keep failing? The exact _why_ is still a mystery to me,

Not that it sounds important anymore to discover the real underlying reason, but throwing out a potential investigative lead in case anything similar pops up later: remember a similar issue with Cirrus CI having specific TCP blackhole settings in its FreeBSD image and had to disable it since otherwise some tests hang-until-killed:

https://github.com/zeek/broker/blob/28fbb63d06c9192923effc930a4b60226c35fb0e/ci/freebsd/prepare.sh#L11-L16

Maybe there's related settings (or something changed in that setting) that caused the behavior.

jsiwek reviewed (commented):
Great work, my opinion is this could still be on track to merge for Zeek 4.0.  There's still some TODOs on the Broker-side of the PR that may take a bit and we don't have practical testing of the changes, but we might rely on the beta-testing phase to take however long it needs to iron out problems/bugs.  The one major thing to resolve on this Zeek-side of the PR is deciding what to do regarding the default routing/forwarding behavior (explained more in the specific review comment later).

> * The `brokerstore-backend-simple-reverse` test relied on a particular ordering of events: clone 1 assumed that it would see `announce_master` _before_ `node_up`. Intuitively, I’d assume the exact opposite. It makes more sense to me to see the `node_up` for manager 1 before this node announces its store master. This test failed after my other patches (which now always seem to trigger `node_up` _first_), so I’ve updated the test script to be less reliant on a particular order

That seems fine to me: with the `Broker::announce_master` and `Cluster::node_up` events being at separate layers (Broker vs. Cluster frameworks), I also would not assume any ordering relationship.  I do also have some intuition and knowledge of things to practically expect that `Cluster::node_up` might come first in an underlying race... the fact that it typically wasn't before and now is makes me think there's a characteristic of the Broker changes that actually improved latency/performance in this case.

rsmmr commented on src/broker/Manager.cc:
```diff
@@ -167,17 +199,16 @@ void Manager::InitPostScript()
 
 	broker::broker_options options;
 	options.disable_ssl = get_option("Broker::disable_ssl")->AsBool();
-	options.forward = get_option("Broker::forward_messages")->AsBool();
```
I would argue for a desired default of "forwarding on" as I think it'd be good to generally decouple reachability from physical connections: Pub/sub "just works" as long as there's *some* path available. Even for a standard cluster, this will be helpful to avoid requiring the full mesh.

That said, I can see doing this in two steps to avoid surprising: 4.0 leaves it off by default, along a note in NEWS that 4.1 will turn it on. So, that would be (2).

On the question whether to expose this option at all: it'll be good to leave it available, even though I hope most people will never touch it once it's on by default.

Re/ naming: not much of an opinion on whether we should introduce a new option `disable_forwarding` vs reusing the existing `forward`.

rsmmr commented:
To record the current plan here: we'd like to merge this into 4.0 still, but will need more testing ; looking into that. In case it won't make it anymore, we'll merge first thing during the 4.1 cycle. In that case, we could still switch to CAF 0.18 for 4.0 to get the performance gains there and make future maintenance easier.

0xxon commented:
One problem of this branch has come up during testing already. Apparently broker now needs much more compiler resources (as in memory) to be compiled.

The required resourced exceed the resources that are available in any besides the arm64 machines on the OpenSuse build service (and this is with non-parallel builds). This means that we currently cannot build binary packages for i586/armhf/aarch64 (or powerpc - this is typically disabled but I test them from time to time).

This also means that Zeek might generally no longer be buildable on raspberry pi devices.

It would be neat if we could reduce the memory footprint that the compiler requires to build this a bit.

Neverlord commented:
> It would be neat if we could reduce the memory footprint that the compiler requires to build this a bit.

Yep, I'm already working on that.

rsmmr commented:
I'm going to make a call here and move this to 4.1.0, it's just getting too late and I'd rather focus on making the move to CAF 0.18 for now. Let me know if anybody objects and we can discuss.

jsiwek commented:
> I'm going to make a call here and move this to 4.1.0, it's just getting too late and I'd rather focus on making the move to CAF 0.18 for now. Let me know if anybody objects and we can discuss.

I don't object, just remarking that if we hold to the plan of "not significantly changing default behaviors during the the course of an LTS", then message routing may remain disabled by default through the duration of the 5.0.x LTS and not be the default until 6.0.x.

rsmmr commented:
> I don't object, just remarking that if we hold to the plan of "not significantly changing default behaviors during the the course of an LTS", then message routing may remain disabled by default through the duration of the 5.0.x LTS and not be the default until 6.0.x.

I wouldn't phrase our policy quite like that. It depends on how disruptive it is (that's the "significantly" part), and we always have our exception clause of being able to decide case-by-case. Here, I think I would propose to start with off by default for 4.1 and then try on by default for 4.2 and see how that goes. My guess is that very few people will even notice. If that's right, then on by default for 5.0 would seem feasible to me, in particular because the work-around for any trouble is easy: flip a single boolean back.

Neverlord commented:
Converting back to draft status, because we picked some commits from this to `master` already and this needs rebasing now.

Neverlord commented:
Closing for now until the redesign for a stable network protocol went through. :)
